sparse/unit_test/Test_Sparse_gmres.hpp

```diff
@@ -0,0 +1,177 @@
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+
+#include <string>
+#include <stdexcept>
+
+#include "KokkosSparse_CrsMatrix.hpp"
+#include "KokkosKernels_IOUtils.hpp"
+#include "KokkosBlas1_nrm2.hpp"
+#include "KokkosSparse_spmv.hpp"
+#include "KokkosSparse_gmres.hpp"
+
+#include <gtest/gtest.h>
+
+using namespace KokkosSparse;
+using namespace KokkosSparse::Experimental;
+using namespace KokkosKernels;
+using namespace KokkosKernels::Experimental;
+
+namespace Test {
+
+template <typename scalar_t, typename lno_t, typename size_type,
+          typename device>
+void run_test_gmres() {
+  using exe_space = typename device::execution_space;
+  using mem_space = typename device::memory_space;
+  using sp_matrix_type =
+      KokkosSparse::CrsMatrix<scalar_t, lno_t, device, void, size_type>;
+  using KernelHandle = KokkosKernels::Experimental::KokkosKernelsHandle<
+      size_type, lno_t, scalar_t, exe_space, mem_space, mem_space>;
+
+  // Create a diagonally dominant sparse matrix to test:
+  constexpr auto n             = 5000;
+  constexpr auto m             = 15;
+  constexpr auto numRows       = n;
+  constexpr auto numCols       = n;
+  constexpr auto diagDominance = 1;
+  constexpr bool verbose       = false;
+
+  typename sp_matrix_type::non_const_size_type nnz = 10 * numRows;
+  auto A = KokkosSparse::Impl::kk_generate_diagonally_dominant_sparse_matrix<
+      sp_matrix_type>(numRows, numCols, nnz, 0, lno_t(0.01 * numRows),
+                      diagDominance);
+
+  // Make kernel handles
+  KernelHandle kh;
+  kh.create_gmres_handle(n, m);
+  auto gmres_handle = kh.get_gmres_handle();
+  using GMRESHandle =
+      typename std::remove_reference<decltype(*gmres_handle)>::type;
+  using ViewVectorType = typename GMRESHandle::nnz_value_view_t;
+  using float_t        = typename GMRESHandle::float_t;
+
+  // Set initial vectors:
+  ViewVectorType X("X", n);    // Solution and initial guess
+  ViewVectorType Wj("Wj", n);  // For checking residuals at end.
+  ViewVectorType B(Kokkos::view_alloc(Kokkos::WithoutInitializing, "B"),
+                   n);  // right-hand side vec
+
+  if (std::is_same<float_t, float>::value) {
+    // reduce tol for float
+    gmres_handle->set_tol(1e-7);
```

Comment:
Had to significantly turn down tol for the test to work with floats. I assume this is OK.